contrib/babelfishpg_tsql/sql/sys_functions.sql

```diff
@@ -3416,9 +3416,39 @@ LANGUAGE C STABLE PARALLEL SAFE;
 CREATE OR REPLACE FUNCTION sys.is_member(IN role sys.SYSNAME)
 RETURNS INT AS
 $$
-	SELECT sys.is_rolemember_internal(role, NULL);
+BEGIN
+    -- Always return 1 for 'public'
+    IF (role = 'public')
+    THEN RETURN 1;
+    END IF;
+
+    IF EXISTS (SELECT orig_loginname FROM sys.babelfish_authid_login_ext WHERE orig_loginname = role)
+    THEN
+        IF EXISTS (SELECT name FROM sys.login_token WHERE name = role)
+        THEN RETURN 1; -- Return 1 if current session user is a member of role or windows group
+        ELSIF NOT EXISTS (SELECT type FROM sys.login_token WHERE type IN ('WINDOWS LOGIN', 'WINDOWS GROUP')) THEN -- session is not a windows auth session
+            IF (CHARINDEX('\', role) != 0)
+            THEN RETURN NULL;  -- argument is a windows group
+            ELSE RETURN sys.is_rolemember_internal(role, NULL); -- argument is not a windows group
+            END IF;
+        ELSE RETURN 0; -- Return 0 if current session user is not a member of role or windows group
+        END IF;
+    ELSIF EXISTS (SELECT orig_username FROM sys.babelfish_authid_user_ext WHERE orig_username = role)
```

Comment:
What if login/server role and user/database role have same name? In that case this implementation will evaluate login membership which is not correct it should be evaluating database role membership as mentioned in first comment.

Reply:
Apart from that we should be filtering based on database id as well, otherwise we would be considering all the user/database roles created across all the databases which is wrong.

Reply:
sys.user_token only shows entries for current database